components/camel-kubernetes/src/main/java/org/apache/camel/component/kubernetes/customresources/KubernetesCustomResourcesConsumer.java

```diff
@@ -0,0 +1,193 @@
+
+    @Override
+    protected void doStart() throws Exception {
+        super.doStart();
+        executor = getEndpoint().createExecutor();
+
+        customResourcesWatcher = new CustomResourcesConsumerTask();
+        executor.submit(customResourcesWatcher);
+    }
+
+    @Override
+    protected void doStop() throws Exception {
+        super.doStop();
+
+        LOG.debug("Stopping Kubernetes Custom Resources Consumer");
+        if (executor != null) {
+            if (getEndpoint() != null && getEndpoint().getCamelContext() != null) {
+                if (customResourcesWatcher != null) {
+                    customResourcesWatcher.getWatch().close();
+                }
+                getEndpoint().getCamelContext().getExecutorServiceManager().shutdownNow(executor);
+            } else {
+                if (customResourcesWatcher != null) {
+                    customResourcesWatcher.getWatch().close();
+                }
+                executor.shutdownNow();
+            }
+        }
+        executor = null;
+    }
+
+    class CustomResourcesConsumerTask implements Runnable {
+
+        private Watch watch;
+
+        @Override
+        public void run() {
+            RawCustomResourceOperationsImpl w = getEndpoint().getKubernetesClient()
+                    .customResource(getCRDContext(getEndpoint().getKubernetesConfiguration()));
+            if (ObjectHelper.isNotEmpty(getEndpoint().getKubernetesConfiguration().getNamespace())) {
+                LOG.error("namespace is not specified.");
+            }
+            String namespace = getEndpoint().getKubernetesConfiguration().getNamespace();
+            try {
+                w.watch(namespace, new Watcher<String>() {
+
+                    @Override
+                    public void eventReceived(Action action, String resource) {
+                        Exchange exchange = getEndpoint().createExchange();
+                        exchange.getIn().setBody(resource);
+                        exchange.getIn().setHeader(KubernetesConstants.KUBERNETES_CRD_EVENT_ACTION, action);
+                        exchange.getIn().setHeader(KubernetesConstants.KUBERNETES_CRD_EVENT_TIMESTAMP,
+                                System.currentTimeMillis());
+                        try {
+                            processor.process(exchange);
+                        } catch (Exception e) {
+                            getExceptionHandler().handleException("Error during processing", exchange, e);
+                        }
+                    }
+
+                    @Override
+                    public void onClose(KubernetesClientException cause) {
+                        if (cause != null) {
+                            LOG.error(cause.getMessage(), cause);
+                        }
+
+                    }
+                });
+            } catch (Exception e) {
+                LOG.error("Exception in handling githubsource instance change", e);
+            }
+
+            /*
```

Comment:
Is there a reason for commented part?

Reply:
No, I forgot to clean up.  I take it out.  Thanks.